File: Java/LargestPalindromeProduct/LargestPalindromeProduct.java
/**
 * (542ms)
 * Created by Phoenix-Z on 2017/6/18.
 */
public class LargestPalindromeProduct {

    public int largestPalindrome(int n) {
        if(n == 1)
            return 9;
        long upperBound = (int)Math.pow(10, n) - 1, lowerBound = upperBound / 10;
        long maxProduct = upperBound * upperBound, product = upperBound * upperBound;
        long leftHalf = maxProduct / (int)Math.pow(10, n);
        boolean palindromeFound = false;
        while(!palindromeFound) {
            product = createPalindrome(leftHalf);
            for(long i = upperBound; i * i >= product; i--) {
                if(product / i <= lowerBound)
                    break;
                if(product % i == 0) {
                    palindromeFound = true;
                    break;
                }
            }
            leftHalf--;
        }
        return (int) (product % 1337);
    }

    private long createPalindrome(long leftHalf) {
        String palindrome = leftHalf + new StringBuilder(Long.toString(leftHalf)).reverse().toString();
        return Long.parseLong(palindrome);
    }

    public static void main(String[] args) {
        LargestPalindromeProduct test = new LargestPalindromeProduct();
        System.out.println(test.largestPalindrome(2));
    }

}

File: Python/LargestPalindromeProduct/LargestPalindromeProduct.py
class Solution(object):
    # (995ms)
    def largestPalindrome(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1:
            return 9
        upper, lower, palindromeFound= 10 ** n - 1, 10 ** (n-1) -1, False
        left = upper * upper / (10**n)
        while not palindromeFound:
            palindrome = int(str(left) + str(left)[::-1])
            if (len(str(palindrome))%2==0):
                for i in xrange(upper / 11,(lower+1) / 11,-1):
                    if (palindrome / (11*i) > upper):
                        break
                    if ((palindrome % (11*i)) == 0):
                        palindromeFound = True
                        break
            else:
                for i in xrange(upper, lower + 1, -1):
                    if (palindrome / i > upper or i * i < palindrome):
                        break
                    if (palindrome % i == 0):
                        palindromeFound = True
                        break   
            left -= 1
        return palindrome % 1337
